### DISCUSSION TITLE
getting a typeerror: deprecated_search

### DISCUSSION THREAD

**Discussion by tbernhardson** [jertel/elastalert2#647] [Q&A]
Running the latest elastalert2 and Elasticsearch 7.16.1 when running either elastalert-test-rule I get the following error:
`INFO:elastalert:Note: In debug mode, alerts will be logged to console but NOT actually sent.
            To send them but remain verbose, use --verbose instead.
Didn't get any results.
INFO:elastalert:Note: In debug mode, alerts will be logged to console but NOT actually sent.
                To send them but remain verbose, use --verbose instead.
1 rules loaded
INFO:apscheduler.scheduler:Adding job tentatively -- it will be properly scheduled when the scheduler starts
Traceback (most recent call last):
  File "/usr/local/bin/elastalert-test-rule", line 33, in <module>
    sys.exit(load_entry_point('elastalert2==2.2.3', 'console_scripts', 'elastalert-test-rule')())
  File "/usr/local/lib/python3.6/site-packages/elastalert2-2.2.3-py3.6.egg/elastalert/test_rule.py", line 518, in main
    test_instance.run_rule_test()
  File "/usr/local/lib/python3.6/site-packages/elastalert2-2.2.3-py3.6.egg/elastalert/test_rule.py", line 508, in run_rule_test
    self.run_elastalert(rule_yaml, conf)
  File "/usr/local/lib/python3.6/site-packages/elastalert2-2.2.3-py3.6.egg/elastalert/test_rule.py", line 424, in run_elastalert
    client.run_rule(rule, self.endtime, self.starttime)
  File "/usr/local/lib/python3.6/site-packages/elastalert2-2.2.3-py3.6.egg/elastalert/elastalert.py", line 929, in run_rule
    if not self.run_query(rule, rule['starttime'], tmp_endtime):
  File "/usr/local/lib/python3.6/site-packages/elastalert2-2.2.3-py3.6.egg/elastalert/elastalert.py", line 661, in run_query
    data = self.get_hits_terms(rule, start, end, index, rule['query_key'])
  File "/usr/local/lib/python3.6/site-packages/elastalert2-2.2.3-py3.6.egg/elastalert/elastalert.py", line 540, in get_hits_terms
    body=query, size=0, ignore_unavailable=True)
  File "/usr/local/lib/python3.6/site-packages/elasticsearch/client/utils.py", line 301, in _wrapped
    return func(*args, params=params, headers=headers, **kwargs)
TypeError: deprecated_search() got an unexpected keyword argument 'headers'
`
When running elastalert I get the following (somewhat the same) error:
`The rule Silent Log Source - Linux has raised an uncaught exception.

It has been disabled and will be re-enabled when ElastAlert restarts or if the rule config file has been modified.

Traceback (most recent call last):
  File "/home/tbernhar/elastalert2/elastalert/elastalert.py", line 1350, in handle_rule_execution
    num_matches = self.run_rule(rule, endtime, rule.get('initial_starttime'))
  File "/home/tbernhar/elastalert2/elastalert/elastalert.py", line 947, in run_rule
    if not self.run_query(rule, rule['starttime'], endtime):
  File "/home/tbernhar/elastalert2/elastalert/elastalert.py", line 661, in run_query
    data = self.get_hits_terms(rule, start, end, index, rule['query_key'])
  File "/home/tbernhar/elastalert2/elastalert/elastalert.py", line 540, in get_hits_terms
    body=query, size=0, ignore_unavailable=True)
  File "/usr/local/lib/python3.6/site-packages/elasticsearch/client/utils.py", line 301, in _wrapped
    return func(*args, params=params, headers=headers, **kwargs)
TypeError: deprecated_search() got an unexpected keyword argument 'headers'
`
my rule file (somewhat redacted) is:
`# Rule Name
name: Silent Log Source - Linux

# Type of alert
type: flatline

# Index to Search
index: filebeat-*

# The minimum number of events to NOT trigger the alert
threshold: 1

# The time period that must contain less than threshold events
timeframe:
  minutes: 60

# Base on unique values of host.name
query_key: "host.name"

# Do an aggregation query to get counts of documents matching each
# unique query_key
use_terms_query: true

# search for doc_type = _type
# required when use_terms_query = true
doc_type: "_doc"

# A list of Elasticsearch filters used for find events
# These filters are joined with AND and nested in a filtered query
# For more info: http://www.elasticsearch.org/guide/en/elasticsearch/reference/c
urrent/query-dsl.html
filter:
- bool:
   must: [ {term: {host.name: ln*}} ]
-bool:
   must: [ {term: {host.name: ln*}} ]

# The alert to use when a match is found
alert:
  - "email"

# Email address(es) to send alerts to
email:
  - "email@company.com"

`
Is this a problem with my rule file or an compatibility issue between elastalert & elasticsearch?

**Comment by tbernhardson** [comment]
Note - tried to use the "Insert Code" looks like it worked 1 out of 3 times...

**Reply by jertel** [reply]
Is this the first time you've installed ElastAlert2? Or were you previously running it successfully, and then upgraded either ElastAlert 2 or Elasticsearch?

**Reply by tbernhardson** [reply]
Brand new Elastalert2 installation.

**Reply by jertel** [reply]
How did you install it?

**Reply by tbernhardson** [reply]
followed the instructions here: https://elastalert2.readthedocs.io/en/latest/running_elastalert.html#downloading-and-configuring

Then setup a directory /etc/elastalert2 & /etc/elastalert2/rules, put the config.yaml file in /etc/elastalert2 (updated for my elasticsearch instance), and the rule yaml file in /etc/elastalert2/rules.

**Comment by jertel** [comment]
Let's start by getting you upgraded to the required (supported) version of Python 3.10. Once you fix that and re-run ElastAlert 2 let's see what it looks like then.

**Comment by tbernhardson** [comment]
yeah, that's going to be a challenge.  None of the Systems/OS's I currently have access to have Python 3.10 available without installing from source...
how far back would I have to go to get a version of elastalert2 that runs on python 3.6 or 3.9?

**Reply by nsano-rururu** [reply]
@tbernhardson 

Can you execute the following command?
If it's not 7.0.0,
Can you run `pip uninstall elasticsearch` and `pip install elasticsearch==7.0.0`?

```sh
pip list | grep elasticsearch
```

**Reply by nsano-rururu** [reply]
By the way, I use it with python 3.9, but there is no problem.

**Reply by jertel** [reply]
Consider trying the Docker image which will let you compare results, and help pinpoint if it's an environmental/installation issue.

**Comment by tbernhardson** [comment]
Docker is not available on any of the systems with access to my elastic infrastructure..
The system I have been working on with elastalert2 is using the python elasticsearch module 7.15.2 (and has other applications that require it).
I will work on moving elastalert to another system where I can use the python3 elasticsearch 7.0.0 module and see if that fixes the issue.

**Reply by tbernhardson** [reply]
using the elasticsearch 7.0.0 module fixed the issue - well, at least the elastalert-test-rule command works now, havn't tried the app itself yet.
